pkg/process/monitor/process_monitor.go

```diff
@@ -119,282 +100,239 @@ type ProcessCallback struct {
 //
-//	2/ Exec callback with a Metadata (!=ANY) must be registered before the sibling Exit metadata,
-//	   otherwise the Subscribe() will return an error as no metadata will be saved between Exec and Exit,
-//	   please refer to GetProcessMonitor()
-func (pm *ProcessMonitor) Subscribe(callback *ProcessCallback) (UnSubscribe func(), err error) {
-	pm.m.Lock()
-	defer pm.m.Unlock()
-
-	for _, c := range pm.procEventCallbacks[callback.Event] {
-		if c == callback {
-			return nil, errors.New("same callback can't be registered twice")
-		}
-	}
+// A callback can be registered only once, callback with a filter type (not ANY) must be registered before the matching
+// Exit callback.
+func (pm *ProcessMonitor) SubscribeExec(callback *ProcessCallback) (func(), error) {
+	pm.processExecCallbacksMutex.Lock()
+	pm.processExecCallbacks[callback] = struct{}{}
+	pm.processExecCallbacksMutex.Unlock()
 
-	// check if the sibling Exec callback exist
-	if callback.Event == EXIT && callback.Metadata != ANY {
-		foundSibling := false
-		for _, c := range pm.procEventCallbacks[EXEC] {
-			if c.Metadata == callback.Metadata && c.Regex.String() == callback.Regex.String() {
-				foundSibling = true
-				break
-			}
-		}
-		if !foundSibling {
-			return nil, errors.New("no Exec callback has been found with the same Metadata and Regex, please Subscribe(Exec callback, Metadata) first")
-		}
-	}
+	// UnSubscribe()
+	return func() {
+		pm.processExecCallbacksMutex.Lock()
+		delete(pm.processExecCallbacks, callback)
+		pm.processExecCallbacksMutex.Unlock()
+	}, nil
+}
 
-	pm.procEventCallbacks[callback.Event] = append(pm.procEventCallbacks[callback.Event], callback)
+// SubscribeExit register an exit callback and returns unsubscribe function callback that removes the callback.
+func (pm *ProcessMonitor) SubscribeExit(callback *ProcessCallback) (func(), error) {
+	pm.processExitCallbacksMutex.Lock()
+	pm.processExitCallbacks[callback] = struct{}{}
+	pm.processExitCallbacksMutex.Unlock()
 
 	// UnSubscribe()
 	return func() {
-		pm.m.Lock()
-		defer pm.m.Unlock()
-
-		// we are scanning all callbacks remove the one we registered
-		// and remove it from the pm.procEventCallbacks[callback.Event] list
-		for i, c := range pm.procEventCallbacks[callback.Event] {
-			if c == callback {
-				l := len(pm.procEventCallbacks[callback.Event])
-				pm.procEventCallbacks[callback.Event][i] = pm.procEventCallbacks[callback.Event][l-1]
-				pm.procEventCallbacks[callback.Event] = pm.procEventCallbacks[callback.Event][:l-1]
-				return
-			}
-		}
+		pm.processExitCallbacksMutex.Lock()
+		delete(pm.processExitCallbacks, callback)
+		pm.processExitCallbacksMutex.Unlock()
 	}, nil
 }
 
+// Stop decreasing the refcount, and if we reach 0 we terminate the main event loop.
 func (pm *ProcessMonitor) Stop() {
-	pm.m.Lock()
-	if pm.refcount == 0 {
-		pm.m.Unlock()
-		return
-	}
-
-	pm.refcount--
-	if pm.refcount > 0 {
-		pm.m.Unlock()
+	if pm.refcount.Dec() != 0 {
```

Comment:
```
if pm.refcount.Load() == 0 || pm.refcount.Dec() <=0 {
   return
}

Reply:
why? that's not the intention of the code

Reply:
I find it less readable for a new developer, and it might introduce races